include/x509_hierarchy.hpp

```diff
@@ -0,0 +1,273 @@
+
+                    str << ' ' << node.certificate.get_common_name() << std::endl;
+                },
+                root, 1);
+        }
+
+        return str.str();
+    }
+
+public:
+    template <typename function> static void for_each_child(function func, const X509Node& node, int depth = 0) {
+        if (node.children.empty())
+            return;
+
+        for (const auto& child : node.children) {
+            func(child, depth);
+
+            if (!child.children.empty()) {
+                for_each_child(func, child, (depth + 1));
+            }
+        }
+    }
+
+    static X509CertificateHierarchy build_hierarchy(std::vector<X509Wrapper>& certificates) {
+        X509CertificateHierarchy ordered;
+
+        // Search for all self-signed certificates and add them as the roots, and also search for
+        // all owner-less certificates and also add them as root, since they are  not self-signed
+        // but we can't find any owner for them anyway
+        std::for_each(certificates.begin(), certificates.end(), [&](const X509Wrapper& certif) {
+            if (certif.is_selfsigned()) {
+                ordered.hierarchy.push_back({certif, certif.get_certificate_hash_data(), certif, {}});
+            } else {
+                // Search for a possible owner
+                bool has_owner = std::find_if(certificates.begin(), certificates.end(), [&](const X509Wrapper& owner) {
+                                     return certif.is_child(owner);
+                                 }) != certificates.end();
+
+                if (!has_owner) {
+                    // If we don't have an owner we can't determine the proper hash, use invalid
+                    CertificateHashData invalid;
+                    ordered.hierarchy.push_back({certif, invalid, certif, {}});
+                }
+            }
+        });
+
+        // Remove all root certificates from the provided certificate list
+        auto remove_roots = std::remove_if(certificates.begin(), certificates.end(), [&](const X509Wrapper& certif) {
+            return std::find_if(ordered.hierarchy.begin(), ordered.hierarchy.end(), [&](const X509Node& node) {
+                       return (certif == node.certificate);
+                   }) != ordered.hierarchy.end();
+        });
+
+        certificates.erase(remove_roots, certificates.end());
+
+        // Try build the full hierarchy, we are not assuming any order
+        while (certificates.size()) {
+            // The current certificate that we're testing for
+            auto current = std::move(certificates.back());
+            certificates.pop_back();
+
+            // If we have any roots try and search in the hierarchy the owner
+            if (!ordered.try_add_to_hierarchy(std::move(current))) {
+                // If we couldn't add to the hierarchy move it down in the queue and try again later
+                certificates.insert(certificates.begin(), std::move(current));
+            }
+        }
```

Comment:
```suggestion
        // Try build the full hierarchy, we are not assuming any order
        // This will not get stuck in a loop, because we removed the orphaned certificates already
        // Note: the logic is fairly simple here, but has cubic runtime in the number of certs
        while (certificates.size()) {
            // The current certificate that we're testing for
            auto current = std::move(certificates.back());
            certificates.pop_back();

            // If we have any roots try and search in the hierarchy the owner
            if (!ordered.try_add_to_hierarchy(std::move(current))) {
                // If we couldn't add to the hierarchy move it down in the queue and try again later
                certificates.insert(certificates.begin(), std::move(current));
            }
        }
```